internal/cmd/commands/server/worker_tags_reload_test.go

```diff
@@ -104,29 +110,88 @@ func TestServer_ReloadWorkerTags(t *testing.T) {
 		w, err := serversRepo.LookupWorkerByName(testController.Context(), name)
 		require.NoError(err)
 		require.NotNil(w)
-		v, ok := w.CanonicalTags()[key]
-		require.True(ok)
+		tags := w.CanonicalTags()
+		v, ok := tags[key]
+		require.True(ok, fmt.Sprintf("EXPECTED: map[%s:%s], ACTUAL: %+v", key, values, tags))
 		require.ElementsMatch(values, v)
 	}
 
 	// Give time to populate up to the controller
-	time.Sleep(10 * time.Second)
+	// Wait until the worker has connected to the controller as seen via
+	// two status updates
+	timeout := time.NewTimer(15 * time.Second)
+	poll := time.NewTimer(0)
+	var w *server.Worker
+	var lastStatusTime time.Time
+	serversRepo, err := testController.Controller().ServersRepoFn()
+	require.NoError(err)
+pollController:
+	for {
+		select {
+		case <-timeout.C:
+			t.Fatalf("timeout waiting for worker to connect to controller")
+		case <-poll.C:
+			w, err = serversRepo.LookupWorkerByName(testController.Context(), "test")
+			require.NoError(err)
+			if w != nil {
+				switch {
+				case lastStatusTime.IsZero():
+					lastStatusTime = w.GetLastStatusTime().AsTime().Round(time.Second)
+				default:
+					if !lastStatusTime.Equal(w.GetLastStatusTime().AsTime().Round(time.Second)) {
+						timeout.Stop()
+						break pollController
+					}
+				}
+			}
+			poll.Reset(time.Second)
+		}
+	}
+
 	fetchWorkerTags("test", "type", []string{"dev", "local"})
 
+	// Set new tags on worker
 	cmd.presetConfig.Store(fmt.Sprintf(workerBaseConfig+tag2Config, key, testController.ClusterAddrs()[0]))
-
 	cmd.SighupCh <- struct{}{}
 	select {
 	case <-cmd.reloadedCh:
 	case <-time.After(15 * time.Second):
-		t.Fatalf("timeout")
+		t.Fatalf("timeout waiting for worker start")
+	}
+
+	// Give time to populate up to the controller
+	timeout.Reset(15 * time.Second)
+	poll.Reset(time.Second)
+	lastStatusTime = time.Time{}
+	startTime := time.Now()
```

Comment:
This is a slightly modified version of the loop. `TestServer_ReloadInitialUpstreams` is different in that it uses a second controller and can set `lastStatusTime` to zero. In this case, the idea is to wait for 2 status updates starting from this point in time.